app/models/event_stream.rb

```diff
@@ -36,7 +36,9 @@ class EventStream < ApplicationRecord
 
   after_commit :emit_notifications, :on => :create
 
-  GROUP_LEVELS = %i(critical detail warning).freeze
+  GROUP_LEVELS = %w(critical detail warning).map do |lvl|
+    [lvl.to_sym, (lvl + '_regex').to_sym]
```

Comment:
I would prefer if we didn't go with 2D array at this point as it's affects code readability and introduces a bunch of annonymous arguments `_` below. So IMAO we should just leave a simple `%i(critical detail warning)` here and update the group-matching `.detect` statement below to append `_regex` suffix on the fly, please see comment below.

Reply:
You are going to have a hard time convincing me that placing a string concatenation inside a loop is a good idea, no matter the readability benefits. Especially since we would be doing same three concatenations over and over again.

Reply:
If you encode the regex directly as mentioned in https://github.com/ManageIQ/manageiq/pull/17772#issuecomment-410750232, then I think you don't need to differentiate with an `_regex` special suffix.

Reply:
So, I'm picturing a YAML that looks like

```yaml
:ems:
  :ems_dummy:
    :event_handling:
      :event_groups:
        :addition:
          :critical:
          - dummy_event01_add
          - dummy_event02_add
          :detail:
          - dm_event_create
          - !ruby/regexp /^dummy_[0-9]{2}_add$/
```

Then when you loop over that you would do something like 

```ruby
GROUP_LEVELS.detect { |lvl| lvl.any? { |v| v === event_type } }
```

Note that `===` does a `match?` for  regexps and `==` for Strings, so it's a nice operator that works for both.

Reply:
I will definitely use the serialized version of regex, since this kills two problems I had in one hit: regex validation and avoiding regex construction on each `self.group_and_level` call.

Reply:
As for the loop, I would still keep the two loops separated, since giving explicitly named event types precedence over regular expression matches since this gives us guarantee that regular expressions do not break existing functionality. For example, provider introducing a bad regular expression cannot "recategorise" some other provider's events.

This separation is independent of the place where regular expressions are listed, so we can still add them to the configuration file under existing levels.

Reply:
Right, you can do `strings, regexes = events.partition { |e| e.kind_of?(Regexp }`, and then process those subcategories separately